temporalio/worker.py

```diff
@@ -0,0 +1,1107 @@
+        )
+        self._running_activities: Dict[bytes, _RunningActivity] = {}
+        self._task: Optional[asyncio.Task] = None
+
+    def config(self) -> WorkerConfig:
+        """Config, as a dictionary, used to create this worker.
+
+        Returns:
+            Configuration, shallow-copied.
+        """
+        config = self._config.copy()
+        config["activities"] = dict(config["activities"])
+        return config
+
+    @property
+    def task_queue(self) -> str:
+        """Task queue this worker is on."""
+        return self._config["task_queue"]
+
+    async def __aenter__(self) -> Worker:
+        """Start the worker and return self for use by ``async with``.
+
+        Returns:
+            Self.
+        """
+        self._start()
+        return self
+
+    async def __aexit__(self, *args) -> None:
+        """Same as :py:meth:`shutdown` for use by ``async with``."""
+        await self.shutdown()
+
+    async def run(self) -> None:
+        """Run the worker and wait on it to be shutdown."""
+        await self._start()
+
+    def _start(self) -> asyncio.Task:
+        if self._task:
+            raise RuntimeError("Already started")
+        self._task = asyncio.create_task(
+            asyncio.wait([asyncio.create_task(self._run_activities())])
+        )
+        return self._task
+
+    async def shutdown(self) -> None:
+        """Shutdown the worker and wait until all activities have completed.
+
+        This will initiate a shutdown and optionally wait for a grace period
+        before sending cancels to all activities.
+        """
+        if not self._task:
+            raise RuntimeError("Never started")
+        graceful_timeout = self._config["graceful_shutdown_timeout"]
+        logger.info(
+            f"Beginning worker shutdown, will wait {graceful_timeout} before cancelling workflows/activities"
+        )
+        # Start shutdown of the bridge
+        bridge_shutdown_task = asyncio.create_task(self._bridge_worker.shutdown())
+        # Wait for the poller loop to stop
+        await self._task
+
+        # Collect all activity tasks while telling them we're shutting down
+        activity_tasks: List[asyncio.Task] = []
+        for activity in self._running_activities.values():
+            if activity.task:
+                activity_tasks.append(activity.task)
+                activity.worker_shutdown_event.set()
+        if activity_tasks:
+            # Wait for any still running after graceful timeout
+            _, still_running = await asyncio.wait(
+                activity_tasks, timeout=graceful_timeout.total_seconds()
+            )
+            if still_running:
+                # Cancel all still running
+                if len(still_running) == 1:
+                    logger.info(f"Cancelling 1 activity that is still running")
+                else:
+                    logger.info(
+                        f"Cancelling {len(still_running)} activities that are still running"
```

Comment:
Is that something we do in every SDK on worker stop?

I of course can add it (I can also add how long it was running, which workflows the activities belong to since there may be many of the same name, etc). Can you clarify exactly what contextual info you would like to see?

Reply:
I don't think we do this in every SDK but if we're already touching this I don't see why not.
We could let users know which (activity ID, workflow ID, run ID) tuples are still running, it might prove useful.

Reply:
Ok, I will add an info-level log line per activity in the loop below. I my try to share activity cancel logging, but maybe not due to log level differences.

Reply:
Added log details for specific activity cancelling on worker shutdown